webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/reports/ReportingServiceImpl.scala

```diff
@@ -430,6 +428,29 @@ trait CachedFindRuleNodeStatusReports extends ReportingService with CachedReposi
     }
   }
 
+  def findSystemAndUserRuleCompliances(
+         nodeIds: Set[NodeId]
+       , filterBySystemRules: Set[RuleId]
+       , filterByUserRules  : Set[RuleId]): IOResult[(Map[NodeId, ComplianceLevel], Map[NodeId, ComplianceLevel])] = {
+    val n1 = System.currentTimeMillis
+    for {
+      reports <- checkAndGetCache(nodeIds)
+      n2      <- currentTimeMillis
+      _       <-  ReportLoggerPure.Cache.debug(s"Get node compliance from cache in: ${n2 - n1}ms")
+      userCompliance   = reports.map { case (nodeId, nodeStatusReport: NodeStatusReport) =>
+                                           (nodeId, complianceByRules(nodeStatusReport, filterByUserRules))
+                                      }
+      systemCompliance = reports.map { case (nodeId, nodeStatusReport: NodeStatusReport) =>
+                                             (nodeId, complianceByRules(nodeStatusReport, filterBySystemRules))
+                                      }
+      n3      <- currentTimeMillis
+      _       <-  ReportLoggerPure.Cache.debug(s"Compute compliance on rules for ${nodeIds.size} node from cache in: ${n3 - n2}ms")
```

Comment:
here, there is the unavoidable overhead of computing the compliancelevel now, rather than later down the chain.
It goes from ~400ms to 600ms